### PR TITLE
feat: enhanced in-place update module to support vertical scaling

### LINKED ISSUE #1212: [feature request] In-place udpate support resources
**What would you like to be added**:
The [In-place Pod Vertical Scaling feature](https://github.com/kubernetes/kubernetes/pull/102884) have been merged in k8s upstream, so openKruise in-place update need support resources, related workloads:
- CloneSet
- Advanced StatefulSet
- Advanced DaemonSet
- SidecarSet

### PR DESCRIPTION
**Pull request by LavenderQAQ**
### Ⅰ. Describe what this PR does
Enhanced the in-place update module to enable kruise's workload to modify resources without restarting the pod.

### Ⅱ. Does this pull request fix one issue?

Fixes #1212 

### Ⅲ. Describe how to verify it
Need to kubernetes v1.27 version, and open feature gate `InPlacePodVerticalScaling`. Here is my kind configuration:
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  "InPlacePodVerticalScaling": true
nodes:
- role: control-plane
  image: kindest/node:v1.27.3@sha256:3966ac761ae0136263ffdb6cfd4db23ef8a83cba8a463690e98317add2c9ba72
- role: worker
  image: kindest/node:v1.27.3@sha256:3966ac761ae0136263ffdb6cfd4db23ef8a83cba8a463690e98317add2c9ba72
- role: worker
  image: kindest/node:v1.27.3@sha256:3966ac761ae0136263ffdb6cfd4db23ef8a83cba8a463690e98317add2c9ba72
- role: worker
  image: kindest/node:v1.27.3@sha256:3966ac761ae0136263ffdb6cfd4db23ef8a83cba8a463690e98317add2c9ba72
```

Build a cloneset (or other workload that includes in-place updates) and set updateStrategy to InPlaceIfPossible. Here's an example:
```yaml
apiVersion: apps.kruise.io/v1alpha1
kind: CloneSet
metadata:
  name: nginx-cloneset
spec:
  updateStrategy:
    type: InPlaceIfPossible
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx-1
        image: nginx:1.21.1
        ports:
        - containerPort: 80
        resources:
          limits:
            cpu: "1"
            memory: "256Mi"
          requests:
            cpu: "0.5"
            memory: "128Mi"
```
Change the workload's resources, wait a while, and you'll find that the pod's resources will change, and the pod won't restart.

### Ⅳ. Special notes for reviews
**The control of Status for vertical scaling requires a higher version of the k8s api, which makes this pr need to wait for kruise version upgrade. It cannot be merged at this time because it is not yet complete.**

### CODE REVIEW

**Comment by LavenderQAQ** [comment]
I think the current version has been upgraded to 1.28, so consider reopening this pr?

**Comment by ABNER-1** [comment]
Hello @LavenderQAQ. 
I think we can continue this work.
Reopen this PR.

**Comment by LavenderQAQ** [comment]
I rebased it first, and I need to take some time to sort through the logic and fill in some tests. After all, this pr is a bit old.

File: pkg/util/inplaceupdate/inplace_update_defaults.go [hunk 1]
```diff
@@ -487,3 +511,391 @@ func checkAllContainersHashConsistent(pod *v1.Pod, runtimeContainerMetaSet *apps
...
+
+	inPlaceUpdateStateJSON, _ := json.Marshal(state)
+	pod.Annotations[appspub.InPlaceUpdateStateKey] = string(inPlaceUpdateStateJSON)
+	return pod, nil
+}
+
+// defaultCalculateInPlaceUpdateSpec calculates diff between old and update revisions.
+// If the diff just contains replace operation of spec.containers[x].image, it will returns an UpdateSpec.
+// Otherwise, it returns nil which means can not use in-place update.
+func defaultCalculateInPlaceUpdateSpecWithVerticalUpdate(oldRevision, newRevision *apps.ControllerRevision, opts *UpdateOptions) *UpdateSpec {
```
Comment (ABNER-1) [inline pkg/util/inplaceupdate/inplace_update_defaults.go:660]: I am inclined to abstract the core logic function and call it within defaultCalculateInPlaceUpdateSpec based on feature gate.

Or abstract the calcOption in defaultCalculateInPlaceUpdateSpec and use calcOption to assemble.

The reason is: this feature is an enhancement to the current implementation, not a fork. The current way of implementation has too much repetitive code.

File: pkg/util/inplaceupdate/inplace_update_vertical.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	IsContainerUpdateCompleted(pod *v1.Pod, container *v1.Container, containerStatus *v1.ContainerStatus, lastContainerStatus appspub.InPlaceUpdateContainerStatus) bool
+	// To determine whether the pod has been successfully vertical updated
+	IsPodUpdateCompleted(pod *v1.Pod) bool
+}
+
+var verticalUpdateOperator VerticalUpdateInterface = nil
+
+// To register vertical update operations,
+// you can register different vertical update implementations here
+func registerVerticalUpdate() {
```
Comment (ABNER-1) [inline pkg/util/inplaceupdate/inplace_update_vertical.go:44]: Why not do it in init function?

File: test/e2e/apps/daemonset.go [hunk 3]
```diff
@@ -331,6 +331,12 @@ var _ = SIGDescribe("DaemonSet", func() {
 					v1.ResourceCPU: resource.MustParse("100m"),
 				},
 			}
+			ads.Spec.Template.Spec.Containers[0].Env = []v1.EnvVar{
+				{
+					Name:  "TEST",
+					Value: "",
+				},
+			}
```
Comment (ABNER-1) [inline test/e2e/apps/daemonset.go:339]: I think this modification seems to be unrelated to the pull request.

**Review by ABNER-1 — CHANGES_REQUESTED** [review]
Some methods in VerticalUpdateInterface appear to be unimplemented.

**Comment by ABNER-1** [comment]
I encountered two errors while adding e2e test cases:

1. When modifying the image and increasing resources simultaneously, I encountered a RunContainerError. I have already submitted an issue to the Kubernetes community at https://github.com/kubernetes/kubernetes/issues/127356. 

2. When I scaled the CPU limit of resources from 100m to 200m, I observed that occasionally the status was first changed by kubelet from `100m InProcess` to 200m, and then it was set back to `100m InProcess` by kubelet, and after a while, it finally became 200m.

File: apis/apps/pub/inplace_update.go [hunk 4]
```diff
@@ -62,12 +62,21 @@ type InPlaceUpdateState struct {
...
+	VerticalUpdateOnly bool `json:"verticalUpdateOnly"`
+
 	// NextContainerImages is the containers with lower priority that waiting for in-place update images in next batch.
 	NextContainerImages map[string]string `json:"nextContainerImages,omitempty"`
 
 	// NextContainerRefMetadata is the containers with lower priority that waiting for in-place update labels/annotations in next batch.
 	NextContainerRefMetadata map[string]metav1.ObjectMeta `json:"nextContainerRefMetadata,omitempty"`
 
+	// NextContainerResources is the containers with lower priority that waiting for in-place update resources in next batch.
+	NextContainerResources map[string]v1.ResourceRequirements `json:"nextContainerResources,omitempty"`
```
Comment (LavenderQAQ) [inline apis/apps/pub/inplace_update.go:78]: Is this to defer the resource updating until the next update operation when both the updated image and the updated resources are present?

Comment (ABNER-1) [reply apis/apps/pub/inplace_update.go:78]: When resizing resources for a multi-container pod, if we wish to adjust each container individually, it's necessary to track resource changes for containers in subsequent batches.

File: Makefile [hunk 5]
```diff
@@ -152,9 +152,15 @@ endif
 create-cluster: $(tools/kind)
 	tools/hack/create-cluster.sh
 
+DISABLE_CSI ?= false
```
[the remark this answers is not in the document]
Comment (ABNER-1) [reply Makefile:155]: This is a simple env to accelerate cluster setup for e2e tests.

File: apis/apps/pub/inplace_update.go (hunk 4 above)
Comment (LavenderQAQ) [reply apis/apps/pub/inplace_update.go:78]: Ok, I got it.

File: pkg/controller/cloneset/core/cloneset_core.go [hunk 6]
```diff
@@ -211,6 +213,15 @@ func (c *commonControl) IgnorePodUpdateEvent(oldPod, curPod *v1.Pod) bool {
 			}
 		}
 	}
+	// only inplace resource resize
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlaceWorkloadVerticalScaling) &&
+		len(curPod.Labels) > 0 && appspub.LifecycleStateType(curPod.Labels[appspub.LifecycleStateKey]) != appspub.LifecycleStateNormal {
+		opts := c.GetUpdateOptions()
+		opts = inplaceupdate.SetOptionsDefaults(opts)
+		if err := containersUpdateCompleted(curPod, opts.CheckContainersUpdateCompleted); err == nil {
```
Comment (furykerry) [inline pkg/controller/cloneset/core/cloneset_core.go:221]: is it possible that containersUpdateCompleted get called twice since L210 had already called containersUpdateCompleted? what is the relation between these two code blocks?

File: pkg/util/inplaceupdate/inplace_update_vertical.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+	// Store container infos whether c is empty or not.
+	// It will be used in IsContainerUpdateCompleted to check container resources is updated.
+	// case: pending pod can also be resize resource
+	state.LastContainerStatuses[container.Name] = c
+}
+
+// UpdateResource implements vertical updates by directly modifying the container's resources,
+// conforming to the k8s community standard
+func (v *VerticalUpdate) UpdateContainerResource(container *v1.Container, newResource *v1.ResourceRequirements) {
+	if container == nil || newResource == nil {
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_vertical.go:147]: we can not return directly since it is possible to set or unset container resource for a pod of burstable qos class

File: pkg/util/inplaceupdate/inplace_update_vertical.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+// 1. exist   exist   exist  => compare cspec and cstatus (important)
+// 2. exist   empty   nil    => change qos is forbidden
+// 3. exist   empty   empty  => change qos is forbidden
+// 4. exist   empty   exist  => change qos is forbidden
+// 5. empty   empty   nil    => this container is not inplace-updated by kruise, ignore
+// 6. empty   exist   exist  => this container is not inplace-updated by kruise, ignore
+// 7. empty   exist   exist  => this container is not inplace-updated by kruise, ignore
+func (v *VerticalUpdate) IsContainerUpdateCompleted(pod *v1.Pod, container *v1.Container, containerStatus *v1.ContainerStatus, lastContainerStatus appspub.InPlaceUpdateContainerStatus) bool {
+	// case 5-7
+	if lastContainerStatus.Resources.Limits == nil && lastContainerStatus.Resources.Requests == nil {
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_vertical.go:182]: can we simply check pod.status.resize, if the field is not empty, the update is not completed

File: pkg/util/inplaceupdate/inplace_update_vertical.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+	// for example: /spec/containers/0/resources/limits/cpu
+	words := strings.Split(op.Path, "/")
+	if len(words) != 7 {
+		return fmt.Errorf("invalid resource path: %s", op.Path)
+	}
+	idx, err := strconv.Atoi(words[3])
+	if err != nil || len(oldTemp.Spec.Containers) <= idx {
+		return fmt.Errorf("invalid container index: %s", op.Path)
+	}
+	quantity, err := resource.ParseQuantity(op.Value.(string))
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_vertical.go:83]: how about the following patch
```
  { "op": "remove", "path": "/spec/containers/0/resources/limits/cpu" }
```

it is possible to unset the container resource for one of the container in the pod of burstable qos class

File: pkg/util/inplaceupdate/inplace_update_vertical.go (hunk 8 above)
Comment (ABNER-1) [reply pkg/util/inplaceupdate/inplace_update_vertical.go:183]: When we only resize limits，filed `.status.resize` will not be changed by api-server and kubelet

File: pkg/util/inplaceupdate/inplace_update_vertical.go (hunk 7 above)
Comment (ABNER-1) [reply pkg/util/inplaceupdate/inplace_update_vertical.go:148]: We should recreate pod when try to change qos class.

File: pkg/util/inplaceupdate/inplace_update_vertical.go (hunk 9 above)
Comment (ABNER-1) [reply pkg/util/inplaceupdate/inplace_update_vertical.go:83]: I will test this case.

Comment (ABNER-1) [reply pkg/util/inplaceupdate/inplace_update_vertical.go:84]: There is a releated issue in kubernetes https://github.com/kubernetes/kubernetes/pull/128718.

**Comment by ABNER-1** [comment]
I believe 1.32 will be a milestone for InPlacePodVerticalScaling. Many bugs will be fixed, and we can test these two remaining issues again at that time.

File: apis/apps/pub/inplace_update.go [hunk 10]
```diff
@@ -62,12 +62,21 @@ type InPlaceUpdateState struct {
 	// UpdateEnvFromMetadata indicates there are envs from annotations/labels that should be in-place update.
 	UpdateEnvFromMetadata bool `json:"updateEnvFromMetadata,omitempty"`
 
+	// UpdateResources indicates there are resources that should be in-place update.
+	UpdateResources bool `json:"updateResources,omitempty"`
+
+	// VerticalUpdateOnly indicates there are only vertical update in this revision.
+	VerticalUpdateOnly bool `json:"verticalUpdateOnly"`
```
Comment (furykerry) [inline apis/apps/pub/inplace_update.go:69]: VerticalUpdateOnly -> UpdateImage

File: pkg/util/inplaceupdate/inplace_update_vertical.go [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+	}
+	if !cmp.Equal(container.Resources.Limits, containerStatus.Resources.Limits) ||
+		!cmp.Equal(container.Resources.Requests, containerStatus.Resources.Requests) {
+		return false
+	}
+	return true
+}
+
+// Get the resource status from the pod and synchronize it to state
+func (v *NativeVerticalUpdate) SyncResource(pod *v1.Pod, state *appspub.InPlaceUpdateState, specifiedContainers []string) {
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_vertical.go:147]: is SyncResource really necessary , can we remove it from the interface ?

File: pkg/util/inplaceupdate/inplace_update_vertical.go [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+	if op.Value == nil {
+		return errors.New("json patch value is nil")
+	}
+	quantity, err := resource.ParseQuantity(op.Value.(string))
+	if err != nil {
+		return fmt.Errorf("parse quantity error: %v", err)
+	}
+
+	if !v.CanResourcesResizeInPlace(words[6]) {
+		return fmt.Errorf("disallowed inplace update resouece: %s", words[6])
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_vertical.go:89]: resouece -> resource

File: pkg/util/inplaceupdate/inplace_update_vertical.go [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	appspub "github.com/openkruise/kruise/apis/apps/pub"
+)
+
+// interface for In-place workload vertical scaling
+type VerticalUpdateInterface interface {
+	// ParseResourcePatch validates and applies the resource patch to the UpdateSpec.
+	ParseResourcePatch(op *jsonpatch.Operation, oldTemp *v1.PodTemplateSpec, updateSpec *UpdateSpec) error
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_vertical.go:36]: ParseResourcePatch -> UpdateInplaceUpdateMetadata

File: pkg/controller/cloneset/core/cloneset_core.go [hunk 14]
```diff
@@ -201,14 +203,25 @@ func (c *commonControl) IgnorePodUpdateEvent(oldPod, curPod *v1.Pod) bool {
...
 		opts := c.GetUpdateOptions()
 		opts = inplaceupdate.SetOptionsDefaults(opts)
 		if err := containersUpdateCompleted(curPod, opts.CheckContainersUpdateCompleted); err == nil {
 			if cond := inplaceupdate.GetCondition(curPod); cond == nil || cond.Status != v1.ConditionTrue {
 				return false
 			}
+			// if InPlaceWorkloadVerticalScaling is enabled, we should not ignore the update event of updating pod
+			// for handling only in-place resource resize
+			if utilfeature.DefaultFeatureGate.Enabled(features.InPlaceWorkloadVerticalScaling) {
+				return false
```
Comment (furykerry) [inline pkg/controller/cloneset/core/cloneset_core.go:223]: shall we compare the pod.status.resize between old and new object, and ignore the event if the resize is the same ?

also, this func is not covered by UT, plz add an ut

File: pkg/util/inplaceupdate/inplace_update_defaults.go [hunk 15]
```diff
@@ -135,15 +140,46 @@ func defaultPatchUpdateSpecToPod(pod *v1.Pod, spec *UpdateSpec, state *appspub.I
 			state.NextContainerImages[c.Name] = newImage
 		}
 	}
+
+	containersResourceChanged := sets.NewString()
```
Comment (furykerry) [inline pkg/util/inplaceupdate/inplace_update_defaults.go:144]: it seems that containersResourceChanged is not used at all. consider replace containersResourceChanged with a map from container name to ResourceRequirements. In this way, one can skip UpdateResource in L153, and call UpdateResource in L180 in one shot

File: pkg/controller/cloneset/core/cloneset_core.go (hunk 14 above)
Comment (ABNER-1) [reply pkg/controller/cloneset/core/cloneset_core.go:223]: When we only resize limit, the resize is always nil

Comment (furykerry) [reply pkg/controller/cloneset/core/cloneset_core.go:223]: or just compare the status.resources between old and new object ?

**Review by furykerry — COMMENTED** [review]
/lgtm
